pkg/controller/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -0,0 +1,209 @@
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, config *configuration.Config) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNSTemplateTier{
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nstemplatetier-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource NSTemplateTier
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.NSTemplateTier{}}, &handler.EnqueueRequestForObject{}, predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource TemplateUpdateRequest and requeue the owner NSTemplateTier
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.TemplateUpdateRequest{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &toolchainv1alpha1.NSTemplateTier{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileNSTemplateTier implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNSTemplateTier{}
+
+// ReconcileNSTemplateTier reconciles a NSTemplateTier object
+type ReconcileNSTemplateTier struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a NSTemplateTier object and makes changes based on the state read
+// and what is in the NSTemplateTier.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNSTemplateTier) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	logger.Info("Reconciling NSTemplateTier")
+
+	// Fetch the NSTemplateTier tier
+	tier := &toolchainv1alpha1.NSTemplateTier{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, tier)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		logger.Error(err, "unable to get the current NSTemplateTier tier")
+		return reconcile.Result{}, err
+	}
+	if requeue, err := r.ensureTemplateUpdateRequest(logger, tier); err != nil {
+		log.Error(err, "unable to ensure TemplateRequestUpdate resource after NSTemplateTier changed.")
```

Comment:
I'm a bit concerned with the error message in the status. What if multiple (many!) errors happen during the process? Are we going to fill the `NSTemplateTier` status with each error? And how do we get back to a normal state (`ready=true`) afterwards?

Reply:
How it works:
When there is an error occurred you don't just return it and end the reconcile. You also put it to the status and exit. If there was an error in the status before it will update it with the current one.
If everything is fine then in the end of the reconcile loop you set Ready=true and erase the message. We do it in all controllers which have statuses.
It looks like:
```
reconcile() {
    err:= ensureSomething()
    if err != nil {
        wrapWithStatus(err)
        return err
    }
    err= ensureSomethingElse()
    if err != nil {
        wrapWithStatus(err)
        return err
    }
    setStatusReady() // set Ready=true & erase (error) message
    return nil
}
```